File: backend/routers/connections.py
"""
DB接続設定 CRUD API
"""
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from database import get_db
from models import Connection
from schemas import (
    ConnectionCreate, ConnectionExport, ConnectionExportItem,
    ConnectionImportBody, ConnectionImportResult,
    ConnectionResponse, ConnectionTestResult, ConnectionUpdate,
)
from services.db_client import test_connection
from utils.crypto import encrypt

router = APIRouter()


def _get_or_404(db: Session, conn_id: int) -> Connection:
    conn = db.get(Connection, conn_id)
    if conn is None:
        raise HTTPException(status_code=404, detail=f"接続設定 ID={conn_id} が見つかりません")
    return conn


@router.get("", response_model=list[ConnectionResponse])
def list_connections(db: Session = Depends(get_db)):
    """接続設定一覧を取得"""
    return db.query(Connection).order_by(Connection.name).all()


@router.post("", response_model=ConnectionResponse, status_code=status.HTTP_201_CREATED)
def create_connection(body: ConnectionCreate, db: Session = Depends(get_db)):
    """接続設定を新規作成"""
    conn = Connection(
        name=body.name,
        db_type=body.db_type,
        host=body.host,
        port=body.port,
        username=body.username,
        password_enc=encrypt(body.password),
        schema_name=body.schema_name,
        use_ssh=body.use_ssh,
        ssh_host=body.ssh_host,
        ssh_port=body.ssh_port,
        ssh_username=body.ssh_username,
        ssh_auth_type=body.ssh_auth_type,
        ssh_password_enc=encrypt(body.ssh_password) if body.ssh_password else None,
        ssh_key_path=body.ssh_key_path,
        local_bind_port=body.local_bind_port,
    )
    db.add(conn)
    db.commit()
    db.refresh(conn)
    return conn


@router.get("/export", response_model=ConnectionExport)
def export_connections(db: Session = Depends(get_db)):
    """接続設定一覧をエクスポート（パスワードを除く）"""
    connections = db.query(Connection).order_by(Connection.name).all()
    items = [
        ConnectionExportItem(
            name=c.name,
            db_type=c.db_type,
            host=c.host,
            port=c.port,
            username=c.username,
            schema_name=c.schema_name,
            use_ssh=c.use_ssh,
            ssh_host=c.ssh_host,
            ssh_port=c.ssh_port,
            ssh_username=c.ssh_username,
            ssh_auth_type=c.ssh_auth_type,
            ssh_key_path=c.ssh_key_path,
            local_bind_port=c.local_bind_port,
        )
        for c in connections
    ]
    return ConnectionExport(
        exported_at=datetime.now(timezone.utc).isoformat(),
        connections=items,
    )


@router.post("/import", response_model=ConnectionImportResult)
def import_connections(body: ConnectionImportBody, db: Session = Depends(get_db)):
    """接続設定をインポート（同名は重複スキップ）"""
    existing_names = {c.name.lower() for c in db.query(Connection).all()}
    created = 0
    skipped = 0
    skipped_names: list[str] = []

    for item in body.connections:
        if item.name.lower() in existing_names:
            skipped += 1
            skipped_names.append(item.name)
            continue
        conn = Connection(
            name=item.name,
            db_type=item.db_type,
            host=item.host,
            port=item.port,
            username=item.username,
            password_enc=encrypt(""),
            schema_name=item.schema_name,
            use_ssh=item.use_ssh,
            ssh_host=item.ssh_host,
            ssh_port=item.ssh_port,
            ssh_username=item.ssh_username,
            ssh_auth_type=item.ssh_auth_type,
            ssh_key_path=item.ssh_key_path,
            local_bind_port=item.local_bind_port,
        )
        db.add(conn)
        existing_names.add(item.name.lower())
        created += 1

    db.commit()
    return ConnectionImportResult(created=created, skipped=skipped, skipped_names=skipped_names)


@router.get("/{conn_id}", response_model=ConnectionResponse)
def get_connection(conn_id: int, db: Session = Depends(get_db)):
    """接続設定を取得"""
    return _get_or_404(db, conn_id)


@router.put("/{conn_id}", response_model=ConnectionResponse)
def update_connection(conn_id: int, body: ConnectionUpdate, db: Session = Depends(get_db)):
    """接続設定を更新"""
    conn = _get_or_404(db, conn_id)

    conn.name = body.name
    conn.db_type = body.db_type
    conn.host = body.host
    conn.port = body.port
    conn.username = body.username
    conn.schema_name = body.schema_name
    conn.use_ssh = body.use_ssh
    conn.ssh_host = body.ssh_host
    conn.ssh_port = body.ssh_port
    conn.ssh_username = body.ssh_username
    conn.ssh_auth_type = body.ssh_auth_type
    conn.ssh_key_path = body.ssh_key_path
    conn.local_bind_port = body.local_bind_port

    # パスワードは入力があった場合のみ更新
    if body.password is not None:
        conn.password_enc = encrypt(body.password)
    if body.ssh_password is not None:
        conn.ssh_password_enc = encrypt(body.ssh_password)

    db.commit()
    db.refresh(conn)
    return conn


@router.delete("/{conn_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_connection(conn_id: int, db: Session = Depends(get_db)):
    """接続設定を削除"""
    conn = _get_or_404(db, conn_id)
    db.delete(conn)
    db.commit()


@router.post("/{conn_id}/test", response_model=ConnectionTestResult)
def test_connection_endpoint(conn_id: int, db: Session = Depends(get_db)):
    """接続テストを実行"""
    conn = _get_or_404(db, conn_id)
    success, message, latency_ms = test_connection(conn)
    return ConnectionTestResult(success=success, message=message, latency_ms=latency_ms)

File: frontend/src/api/connections.ts
import apiClient from './client'
import type { Connection, ConnectionExport, ConnectionExportItem, ConnectionFormValues, ConnectionImportResult, ConnectionTestResult, DbType } from '@/types'

// バックエンドのスネークケース ↔ フロントエンドのキャメルケース変換
function toSnakeCase(data: ConnectionFormValues) {
  return {
    name: data.name,
    db_type: data.dbType,
    host: data.host,
    port: data.port,
    username: data.username,
    password: data.password,
    schema_name: data.schemaName,
    use_ssh: data.useSsh,
    ssh_host: data.sshHost,
    ssh_port: data.sshPort,
    ssh_username: data.sshUsername,
    ssh_auth_type: data.sshAuthType,
    ssh_password: data.sshPassword,
    ssh_key_path: data.sshKeyPath,
    local_bind_port: data.localBindPort,
  }
}

function toCamelCase(data: Record<string, unknown>): Connection {
  return {
    id: data.id as number,
    name: data.name as string,
    dbType: (data.db_type as DbType) ?? 'mysql',
    host: data.host as string,
    port: data.port as number,
    username: data.username as string,
    schemaName: data.schema_name as string,
    useSsh: data.use_ssh as boolean,
    ssh: data.use_ssh ? {
      host: data.ssh_host as string,
      port: data.ssh_port as number,
      username: data.ssh_username as string,
      authType: data.ssh_auth_type as 'password' | 'key',
      keyPath: data.ssh_key_path as string | undefined,
      localBindPort: data.local_bind_port as number | undefined,
    } : undefined,
    createdAt: data.created_at as string,
    updatedAt: data.updated_at as string,
  }
}

export const connectionsApi = {
  /** 接続設定一覧を取得 */
  list: async (): Promise<Connection[]> => {
    const data = await apiClient.get<Record<string, unknown>[]>('/connections')
    return data.map(toCamelCase)
  },

  /** 接続設定を取得 */
  get: async (id: number): Promise<Connection> => {
    const data = await apiClient.get<Record<string, unknown>>(`/connections/${id}`)
    return toCamelCase(data)
  },

  /** 接続設定を新規作成 */
  create: async (values: ConnectionFormValues): Promise<Connection> => {
    const data = await apiClient.post<Record<string, unknown>>('/connections', toSnakeCase(values))
    return toCamelCase(data)
  },

  /** 接続設定を更新 */
  update: async (id: number, values: ConnectionFormValues): Promise<Connection> => {
    const data = await apiClient.put<Record<string, unknown>>(`/connections/${id}`, toSnakeCase(values))
    return toCamelCase(data)
  },

  /** 接続設定を削除 */
  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/connections/${id}`)
  },

  /** 接続テスト */
  test: async (id: number): Promise<ConnectionTestResult> => {
    return apiClient.post<ConnectionTestResult>(`/connections/${id}/test`)
  },

  /** 接続設定をエクスポート（パスワードを除く JSON を返す） */
  export: async (): Promise<ConnectionExport> => {
    return apiClient.get<ConnectionExport>('/connections/export')
  },

  /** 接続設定をインポート */
  import: async (connections: ConnectionExportItem[]): Promise<ConnectionImportResult> => {
    return apiClient.post<ConnectionImportResult>('/connections/import', { connections })
  },
}
